pkg/services/system_intakes.go

```diff
@@ -427,3 +427,65 @@ func NewUpdateRejectionFields(
 		return updated, nil
 	}
 }
+
+// NewProvideGRTFeedback gives a function to provide GRT Feedback to an intake
+func NewProvideGRTFeedback(
+	config Config,
+	fetch func(c context.Context, id uuid.UUID) (*models.SystemIntake, error),
+	update func(context.Context, *models.SystemIntake) (*models.SystemIntake, error),
+	saveAction func(context.Context, *models.Action) error,
+	saveGRTFeedback func(context.Context, *models.GRTFeedback) (*models.GRTFeedback, error),
+	fetchUserInfo func(context.Context, string) (*models.UserInfo, error),
+	sendReviewEmail func(ctx context.Context, emailText string, recipientAddress string, intakeID uuid.UUID) error,
+) func(context.Context, *models.GRTFeedback, *models.Action, models.SystemIntakeStatus) (*models.GRTFeedback, error) {
+	return func(ctx context.Context, grtFeedback *models.GRTFeedback, action *models.Action, newStatus models.SystemIntakeStatus) (*models.GRTFeedback, error) {
+		intake, err := fetch(ctx, grtFeedback.IntakeID)
+		if err != nil {
+			return nil, err
+		}
+
+		requesterInfo, err := fetchUserInfo(ctx, intake.EUAUserID.ValueOrZero())
+		if err != nil {
+			return nil, err
+		}
+		if requesterInfo == nil || requesterInfo.Email == "" {
+			return nil, &apperrors.ExternalAPIError{
+				Err:       errors.New("requester info fetch was not successful when submitting GRT Feedback"),
+				Model:     intake,
+				ModelID:   intake.ID.String(),
+				Operation: apperrors.Fetch,
+				Source:    "CEDAR LDAP",
+			}
+		}
+
+		if grtFeedback, err = saveGRTFeedback(ctx, grtFeedback); err != nil {
+			return nil, err
+		}
+
+		if err = saveAction(ctx, action); err != nil {
+			return nil, err
+		}
+
+		// we only want to bring over the fields specifically
+		// dealing with Rejection information
+		updatedTime := config.clock.Now()
+		intake.UpdatedAt = &updatedTime
+		intake.Status = newStatus
+		intake, err = update(ctx, intake)
```

Comment:
In a lot of the actions, we only have a few system intake fields we want to update. By fetching the full intake, we reduce the params we have to pass into the mutation, but the tradeoff is we create the potential for race conditions. Since we have enough actions, what do you think about creating a storage function that just updates the system intake status? Looks like it would take three params: IntakeID, NewStatus, and UpdatedAt.

Reply:
I think that's a great idea, but I'd rather do it in a followup. I've found a good amount of small followups I want to do, so I'll add this to that list.